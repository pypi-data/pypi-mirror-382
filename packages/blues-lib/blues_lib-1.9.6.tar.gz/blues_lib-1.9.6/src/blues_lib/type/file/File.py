import sys,os,re,time


from blues_lib.config.ConfigManager import config
from blues_lib.util.BluesURL import BluesURL   
from blues_lib.util.BluesFiler import BluesFiler    

class File():

  ROOT_DIR = config.get("app.rootdir")

  @classmethod
  def get_dir_path(cls,subdir=None):
    '''
    Get the blues lib's standard dir path
    Parameter:
      subdir {None|str|list} : the sub dirs
        None: return the root dir
        str: return the subdir
        list: return the multi subdir
    '''
    if not subdir:
      dir_path = cls.ROOT_DIR 
    else:
      dirs = subdir if type(subdir)==list else [subdir]
      # support multi level dirs
      dir_path = os.path.join(cls.ROOT_DIR,*dirs)

    # create dir
    BluesFiler.makedirs(dir_path)
    return dir_path 

  @classmethod
  def get_file_path(cls,subdir,filename):
    '''
    Get the file absolute path base the blues lib root dir
    Support add subdirs
    Parameter:
      subdir {str,list} : one or multi level dirs
      filename {str} : the filename
    '''
    dir_path = cls.get_dir_path(subdir)
    return os.path.join(dir_path,filename)

  @classmethod
  def get_domain_file_path(cls,url,subdir,extension='txt'):
    domain = BluesURL.get_main_domain(url)
    filename = '%s.%s' % (domain,extension)
    return cls.get_file_path(subdir,filename)

  @classmethod
  def get_file_name(cls,prefix='',name='',suffix='',extension=''):
    file_name = name if name else int(time.time()*1000)
    if prefix:
      file_name='%s-%s' % (prefix,file_name)
    if suffix:
      file_name='%s-%s' % (file_name,suffix)
    if extension:
      file_name='%s.%s' % (file_name,extension)
    return file_name 

