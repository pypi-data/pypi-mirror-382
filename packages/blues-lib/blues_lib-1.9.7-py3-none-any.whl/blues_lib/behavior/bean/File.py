import sys,os,re
from typing import Any

from blues_lib.behavior.Bean import Bean

class File(Bean):

  def _get(self)->Any:
    kwargs = self._get_kwargs(['target_CS_WE','parent_CS_WE','timeout'])
    return self._browser.element.info.get_value(**kwargs)

  def _set(self)->Any:
    kwargs = self._get_kwargs(['target_CS_WE','value','wait_time','parent_CS_WE','timeout'])
    if self._to_be_presence():
      return self._browser.element.file.write(**kwargs)
