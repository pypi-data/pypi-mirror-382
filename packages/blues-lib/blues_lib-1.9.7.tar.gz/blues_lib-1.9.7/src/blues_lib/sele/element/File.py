import sys,os,re,time
from .deco.InfoKeyDeco import InfoKeyDeco

from blues_lib.sele.waiter.Querier import Querier  
from blues_lib.sele.element.Info import Info  
from blues_lib.util.BluesFiler import BluesFiler
from blues_lib.util.BluesDateTime import BluesDateTime

class File():

  def __init__(self,driver):
    self.__driver = driver
    self.__querier = Querier(driver,5) 
    self.__info = Info(driver) 

  def write(self,target_CS_WE,value,wait_time=3,parent_CS_WE=None,timeout=5):
    '''
    Add one or multiple files to the file input
    If there are multiple files, the upload mode is controlled based on whether multiple file upload is supported
    '''

    files = value if type(value) == list else [value]
    # Supports uploading multiple images at a time
    exist_files = BluesFiler.filter_exists(files)
    web_element = self.__querier.query(target_CS_WE,parent_CS_WE,timeout)
    if not exist_files or not web_element:
      return

    self._set_element_visible(web_element)

    is_multiple = self.__info.get_attr(web_element,'multiple')
    if is_multiple:
      # must join the file paths by \n
      file_lines = '\n'.join(exist_files)
      web_element.send_keys(file_lines)
      BluesDateTime.count_down({'duration':wait_time,'title':'Wait image upload...'})
    else:
      for exist_file in exist_files:
        web_element.send_keys(exist_file)
        BluesDateTime.count_down({'duration':wait_time,'title':'Wait image upload...'})
        
  def _set_element_visible(self,web_element):
    """
    确保文件输入框可见且具备可交互尺寸
    - 若元素隐藏(display: none)，设置为inline
    - 若元素尺寸为0(width/height=0)，设置为auto
    """
    driver = self.__driver
    
    # 获取当前样式属性
    current_display:bool = web_element.is_displayed()
    # 获取元素尺寸（rect返回字典包含width/height）
    element_rect:dict = web_element.rect
    current_width = element_rect['width']
    current_height = element_rect['height']
    
    # 调整display属性（如果隐藏）
    if not current_display:
      driver.execute_script("arguments[0].style.display = 'inline';", web_element)
    
    # 调整尺寸（如果为0或接近0）
    min_size = 1  # 最小有效尺寸阈值（像素）
    if current_width <= min_size:
      driver.execute_script("arguments[0].style.width = 'auto';", web_element)
    if current_height <= min_size:
      driver.execute_script("arguments[0].style.height = 'auto';", web_element)
