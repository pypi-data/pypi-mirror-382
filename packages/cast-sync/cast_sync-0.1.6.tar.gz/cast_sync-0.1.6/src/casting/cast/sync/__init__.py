"""Cast Sync - 3-way sync engine and conflict handling."""

from casting.cast.sync.conflict import ConflictResolution, handle_conflict
from casting.cast.sync.hsync import HorizontalSync, SyncDecision, SyncPlan
from casting.cast.sync.index import build_ephemeral_index
from casting.cast.sync.cbsync import CodebaseSync

__all__ = [
    "HorizontalSync",
    "SyncDecision",
    "SyncPlan",
    "ConflictResolution",
    "handle_conflict",
    "build_ephemeral_index",
    "CodebaseSync",
]

__version__ = "0.2.2"
