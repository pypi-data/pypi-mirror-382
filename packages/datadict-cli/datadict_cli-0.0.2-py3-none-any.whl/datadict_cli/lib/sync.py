"""
Metadata synchronization module for DataDict.

This module handles synchronization between remote metadata sources (databases, APIs, etc.)
and local DataDict catalogs. It compares metadata states and generates ordered change operations
to bring local data in sync with remote sources.

Key Design Principles:
- Source-agnostic: Works with any remote metadata source
- Side-effect free: Functions only compute changes, don't apply them
- Hierarchical-aware: Handles parent-child relationships correctly
- Filesystem-independent: Change application is handled elsewhere

Data Format:
Both remote and local data are provided as pandas DataFrames with these columns:
- type: Item type (database, schema, table, column)
- name: Item name
- key: Fully qualified name (e.g., "database.schema.table")
- sub_type: Optional subtype classification
- data_type: Data type information (for columns)
- parent_key: Parent's fully qualified name (null for root items)

Local data includes an additional column:
- archived: Boolean flag indicating if item is archived

Change Ordering Rules:
1. Changes are returned in four separate arrays: create, modify, archive, unarchive
2. Each array is ordered by tree depth (root items first, then children)
3. Reparenting operations require archiving all children of the moved subtree
4. Archived items missing from remote are ignored (no change needed)

Special Cases:
- Reparenting: When parent_key changes, archive entire subtree and recreate under new parent
- Archived items: Items archived locally but missing from remote are left unchanged
"""

import hashlib

import pandas as pd

from .types import ChangeType, ItemChange, SyncChanges


def calculate_metadata_checksum(row: pd.Series, exclude_columns: set[str] | None = None) -> str:
    """
    Calculate checksum for metadata row, excluding specified columns.

    Args:
        row: pandas Series representing a single item's metadata
        exclude_columns: Set of column names to exclude from checksum (default: {'archived'})

    Returns:
        Hex string checksum of the row's metadata
    """
    if exclude_columns is None:
        # Exclude operational/derived columns that shouldn't affect value comparisons
        exclude_columns = {
            "archived",
            "id",
            "parent_id",
            "parent_key",
            "depth",
            "checksum",
            "_merge",
        }

    # Get all columns except excluded ones, sorted for consistency
    metadata_columns = sorted([col for col in row.index if col not in exclude_columns])

    # Create string representation of metadata values
    metadata_values = []
    for col in metadata_columns:
        value = row[col]
        # Handle None/NaN values consistently
        if pd.isna(value) or value is None:
            metadata_values.append("")
        else:
            metadata_values.append(str(value))

    # Join values with separator and create hash
    content = "|".join(metadata_values)
    return hashlib.sha256(content.encode("utf-8")).hexdigest()


def add_checksum_column(df: pd.DataFrame, exclude_columns: set[str] | None = None) -> pd.DataFrame:
    """
    Add checksum column to DataFrame for metadata comparison.

    Args:
        df: DataFrame with item metadata
        exclude_columns: Set of column names to exclude from checksum (default: {'archived'})

    Returns:
        DataFrame with added 'checksum' column
    """
    if df.empty:
        return df.copy()

    df_copy = df.copy()
    df_copy["checksum"] = df_copy.apply(
        lambda row: calculate_metadata_checksum(row, exclude_columns), axis=1
    )
    return df_copy


def add_depth_column(df: pd.DataFrame) -> pd.DataFrame:
    """
    Adds a 'depth' column which contains the index of the depth along with each column
    """
    if df.empty:
        return df.copy()

    result_df = df.copy()
    result_df["depth"] = -1

    # Normalize parent_key column
    if "parent_key" not in result_df.columns:
        result_df["parent_key"] = None
    if "key" not in result_df.columns:
        result_df["key"] = None

    # Roots: items with no parent_key
    root_mask = result_df["parent_key"].isna()
    result_df.loc[root_mask, "depth"] = 0

    # Map key -> depth for assigned nodes
    name_to_depth: dict[str, int] = {}
    for _, row in result_df[result_df["depth"] == 0].iterrows():
        fn = row.get("key")
        if isinstance(fn, str) and fn:
            name_to_depth[fn] = 0

    current_depth = 0
    while True:
        # Collect names at current depth
        names_at_current_depth = {k for k, v in name_to_depth.items() if v == current_depth}
        if not names_at_current_depth:
            break

        # Assign next level: parent_key in names_at_current_depth and depth not set
        next_level_mask = (result_df["depth"] == -1) & (
            result_df["parent_key"].isin(list(names_at_current_depth))
        )
        next_level = result_df[next_level_mask]
        if next_level.empty:
            break

        result_df.loc[next_level_mask, "depth"] = current_depth + 1
        for _, row in next_level.iterrows():
            fn = row.get("key")
            if isinstance(fn, str) and fn:
                name_to_depth[fn] = current_depth + 1

        current_depth += 1

    # Any remaining unassigned nodes (disconnected) get lowest priority depth
    result_df.loc[result_df["depth"] == -1, "depth"] = current_depth + 1

    return result_df


def extract_change_dataframes(remote_df: pd.DataFrame, local_df: pd.DataFrame) -> dict:
    """
    Extract different types of changes by merging remote and local dataframes.
    Returns a dictionary with modify, create, and archive dataframes.
    """
    remote_with_checksum = add_checksum_column(remote_df)
    local_with_checksum = add_checksum_column(local_df)

    merged = remote_with_checksum.merge(
        local_with_checksum,
        left_on="key",
        right_on="key",
        how="outer",
        indicator=True,
        suffixes=("_remote", "_local"),
    )

    create = merged[merged["_merge"] == "left_only"]
    archive = merged[merged["_merge"] == "right_only"]

    both_df = merged[merged["_merge"] == "both"]
    if both_df.empty:
        modify = both_df
    else:
        modify = both_df[both_df["checksum_remote"] != both_df["checksum_local"]]

    return {"modify": modify, "create": create, "archive": archive}


def _safe_get_value(row: pd.Series, key: str) -> str | None:
    """
    Safely extract value from pandas row, handling NaN values.

    Args:
        row: pandas Series
        key: column key to extract

    Returns:
        String value or None (never NaN)
    """
    value = row.get(key)
    if pd.isna(value) or value is None:
        return None
    return str(value) if not isinstance(value, str) else value


def get_changes(remote_df: pd.DataFrame, local_df: pd.DataFrame) -> SyncChanges:
    """
    Compare remote and local metadata DataFrames and generate ordered change operations.

    Args:
        remote_df: DataFrame with remote metadata (source of truth)
        local_df: DataFrame with local metadata (includes 'archived' column)

    Returns:
        SyncChanges object with four arrays of changes ordered by tree depth
    """
    # Normalize inputs (do not mutate args)
    remote = remote_df.copy()
    local = local_df.copy()

    # Ensure archived column exists on local for consistent logic
    if "archived" not in local.columns:
        local["archived"] = False

    # Ensure required columns exist for depth computation
    if "parent_key" not in remote.columns:
        remote["parent_key"] = None
    if "parent_key" not in local.columns:
        local["parent_key"] = None

    # Ensure key column exists
    if "key" not in remote.columns:
        remote["key"] = None
    if "key" not in local.columns:
        local["key"] = None

    # Compute per-tree depths to order operations correctly
    remote_with_depth = add_depth_column(remote)
    local_with_depth = add_depth_column(local)

    remote_depth_by_name: dict[str, int] = (
        {
            str(row["key"]): int(row["depth"])  # type: ignore[index]
            for _, row in remote_with_depth.iterrows()
            if isinstance(row.get("key"), str)
        }
        if not remote_with_depth.empty
        else {}
    )
    local_depth_by_name: dict[str, int] = (
        {
            str(row["key"]): int(row["depth"])  # type: ignore[index]
            for _, row in local_with_depth.iterrows()
            if isinstance(row.get("key"), str)
        }
        if not local_with_depth.empty
        else {}
    )

    # Merge to compute differences and access both sides' columns
    remote_wc = add_checksum_column(remote)
    local_wc = add_checksum_column(local)

    merged = remote_wc.merge(
        local_wc,
        on="key",
        how="outer",
        indicator=True,
        suffixes=("_remote", "_local"),
    )

    # Identify sets
    create_df = merged[merged["_merge"] == "left_only"]
    archive_df = merged[merged["_merge"] == "right_only"]

    both_df = merged[merged["_merge"] == "both"]
    # Prepare archived series as boolean for safe logical operations
    if "archived" in both_df.columns:
        archived_column = both_df["archived"].infer_objects(copy=False)
        archived_series = archived_column.fillna(False).astype(bool)
    else:
        archived_series = pd.Series(False, index=both_df.index)

    # Only consider modify for items not archived locally to avoid duplicate modify+unarchive
    if both_df.empty:
        modify_df = both_df
        unarchive_df = both_df
    else:
        checksum_diff = both_df["checksum_remote"] != both_df["checksum_local"]
        modify_df = both_df[checksum_diff & (~archived_series)]
        # Unarchive: present in both, archived locally
        unarchive_df = both_df[archived_series]

    # Build ItemChange lists
    create_changes: list[ItemChange] = []
    for _, row in create_df.iterrows():
        depth = remote_depth_by_name.get(_safe_get_value(row, "key"), 0)
        create_changes.append(
            ItemChange(
                change=ChangeType.CREATE,
                type=_safe_get_value(row, "type_remote"),
                name=_safe_get_value(row, "name_remote"),
                key=_safe_get_value(row, "key"),
                sub_type=_safe_get_value(row, "sub_type_remote"),
                data_type=_safe_get_value(row, "data_type_remote"),
                parent_key=_safe_get_value(row, "parent_key_remote"),
                depth=depth,
            )
        )

    modify_changes: list[ItemChange] = []
    for _, row in modify_df.iterrows():
        depth = remote_depth_by_name.get(_safe_get_value(row, "key"), 0)
        modify_changes.append(
            ItemChange(
                change=ChangeType.MODIFY,
                type=_safe_get_value(row, "type_remote"),
                name=_safe_get_value(row, "name_remote"),
                key=_safe_get_value(row, "key"),
                sub_type=_safe_get_value(row, "sub_type_remote"),
                data_type=_safe_get_value(row, "data_type_remote"),
                parent_key=_safe_get_value(row, "parent_key_remote"),
                depth=depth,
            )
        )

    archive_changes: list[ItemChange] = []
    for _, row in archive_df.iterrows():
        # Skip if already archived locally and missing remotely (no-op per rules)
        archived_val = row.get("archived", False)
        try:
            is_archived = False if pd.isna(archived_val) else bool(archived_val)
        except Exception:
            is_archived = False
        if is_archived:
            continue
        depth = local_depth_by_name.get(_safe_get_value(row, "key"), 0)
        archive_changes.append(
            ItemChange(
                change=ChangeType.ARCHIVE,
                type=_safe_get_value(row, "type_local"),
                name=_safe_get_value(row, "name_local"),
                key=_safe_get_value(row, "key"),
                sub_type=_safe_get_value(row, "sub_type_local"),
                data_type=_safe_get_value(row, "data_type_local"),
                parent_key=_safe_get_value(row, "parent_key_local"),
                depth=depth,
            )
        )

    unarchive_changes: list[ItemChange] = []
    for _, row in unarchive_df.iterrows():
        fn = _safe_get_value(row, "key")
        depth = remote_depth_by_name.get(fn, local_depth_by_name.get(fn, 0))
        unarchive_changes.append(
            ItemChange(
                change=ChangeType.UNARCHIVE,
                type=_safe_get_value(row, "type_remote"),
                name=_safe_get_value(row, "name_remote"),
                key=_safe_get_value(row, "key"),
                sub_type=_safe_get_value(row, "sub_type_remote"),
                data_type=_safe_get_value(row, "data_type_remote"),
                parent_key=_safe_get_value(row, "parent_key_remote"),
                depth=depth,
            )
        )

    # Order by depth (root first)
    create_changes.sort(key=lambda c: (c.depth or 0, c.key or ""))
    modify_changes.sort(key=lambda c: (c.depth or 0, c.key or ""))
    archive_changes.sort(key=lambda c: (c.depth or 0, c.key or ""))
    unarchive_changes.sort(key=lambda c: (c.depth or 0, c.key or ""))

    return SyncChanges(
        create=create_changes,
        modify=modify_changes,
        archive=archive_changes,
        unarchive=unarchive_changes,
    )
