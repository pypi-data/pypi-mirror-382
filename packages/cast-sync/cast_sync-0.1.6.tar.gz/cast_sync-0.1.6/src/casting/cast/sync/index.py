"""Ephemeral index building for casts."""

import logging
from pathlib import Path

from casting.cast.core import (
    compute_digest,
    ensure_cast_fields,
    extract_cast_fields,
    parse_cast_file,
    write_cast_file,
)
from casting.cast.core.models import FileRec
from casting.cast.core.yamlio import parse_hsync_entries, reorder_cast_fields

logger = logging.getLogger(__name__)


class EphemeralIndex:
    """In-memory index of a cast's files."""

    def __init__(self, vault_path: Path):
        self.vault_path = vault_path
        self.by_id: dict[str, FileRec] = {}
        self.by_path: dict[str, str] = {}  # relpath -> id

    def add_file(self, rec: FileRec) -> None:
        """Add a file record to the index."""
        self.by_id[rec["id"]] = rec
        self.by_path[rec["relpath"]] = rec["id"]

    def get_by_id(self, identifier: str) -> FileRec | None:
        """Get file record by id."""
        return self.by_id.get(identifier)

    def get_by_path(self, relpath: str) -> FileRec | None:
        """Get file record by relative path."""
        identifier = self.by_path.get(relpath)
        return self.by_id.get(identifier) if identifier else None

    def all_peers(self) -> set[str]:
        """Get all unique peer names referenced."""
        peers = set()
        for rec in self.by_id.values():
            peers.update(rec["peers"].keys())
        return peers

    def all_codebases(self) -> set[str]:
        """Get all unique codebase names referenced."""
        codebases = set()
        for rec in self.by_id.values():
            codebases.update(rec["codebases"])
        return codebases


def build_ephemeral_index(
    root_path: Path, vault_path: Path, fixup: bool = True, limit_file: str | None = None
) -> EphemeralIndex:
    """
    Build an ephemeral index of cast files in a cast folder.

    Args:
        root_path: Path to Cast root (contains .cast/)
        vault_path: Path to cast folder
        fixup: Whether to fix missing id and reorder fields
        limit_file: Optional id or relpath to limit indexing to one file

    Returns:
        EphemeralIndex instance
    """
    index = EphemeralIndex(vault_path)

    # Find all Markdown files
    md_files: list[Path] = []
    if limit_file:
        # Normalize limit_file to a path relative to the cast folder, supporting:
        #  - absolute paths under the cast folder
        #  - callers that include the cast folder prefix (e.g. "Cast/foo.md")
        lf = Path(limit_file)
        candidates: list[Path] = []

        if lf.is_absolute():
            try:
                candidates.append(lf.relative_to(vault_path))
            except ValueError:
                # Not under this cast folder; leave empty so we fall back to id lookup.
                pass
        else:
            if lf.parts and lf.parts[0] == vault_path.name:
                candidates.append(Path(*lf.parts[1:]))
            candidates.append(lf)

        for rel in candidates:
            cand = vault_path / rel
            if cand.exists():
                md_files = [cand]
                break

        if not md_files:
            # Maybe limit_file was an id; scan all and resolve by id below.
            md_files = list(vault_path.rglob("*.md"))
    else:
        md_files = list(vault_path.rglob("*.md"))

    for md_path in md_files:
        try:
            # Parse file
            front_matter, body, has_cast_fields = parse_cast_file(md_path)

            if not has_cast_fields:
                continue

            # Ensure cast fields and canonicalize order (including last-updated first)
            if fixup and front_matter:
                modified = False
                # Legacy migration: cast-vaults -> cast-hsync
                if "cast-vaults" in front_matter and "cast-hsync" not in front_matter:
                    front_matter["cast-hsync"] = front_matter.pop("cast-vaults")
                    modified = True
                front_matter, fields_modified = ensure_cast_fields(front_matter, generate_id=True)
                modified = modified or fields_modified

                # Determine if YAML needs reordering even if no fields were added
                need_reorder = False
                try:
                    keys = list(front_matter.keys())
                    # Enforce 'last-updated' first when present
                    if "last-updated" in front_matter:
                        if not keys or keys[0] != "last-updated":
                            need_reorder = True
                    # Enforce canonical cast-* ordering
                    # Compare current mapping to reorder_cast_fields result
                    # (this catches list canonicalization for 'cast-hsync' and 'cast-codebases')
                    reordered = reorder_cast_fields(dict(front_matter))
                    if list(reordered.keys()) != keys:
                        need_reorder = True
                    else:
                        if reordered.get("cast-hsync") != front_matter.get("cast-hsync"):
                            need_reorder = True
                        if reordered.get("cast-codebases") != front_matter.get("cast-codebases"):
                            need_reorder = True
                except Exception:
                    # If we can't inspect order, be conservative and rewrite
                    need_reorder = True

                if modified or need_reorder:
                    write_cast_file(md_path, front_matter, body, reorder=True)
                    logger.info(f"Fixed cast fields/order in {md_path}")

            identifier = None
            if front_matter:
                identifier = front_matter.get("id")

            if not front_matter or not identifier:
                continue

            # Extract cast fields
            cast_fields = extract_cast_fields(front_matter)
            file_id = cast_fields.get("id") or identifier

            # Parse hsync entries
            hsync_entries = cast_fields.get("cast-hsync") or cast_fields.get("cast-vaults", [])
            peers = parse_hsync_entries(hsync_entries)

            # Get codebases
            codebases = cast_fields.get("cast-codebases", [])
            if not isinstance(codebases, list):
                codebases = []

            # Compute digest
            digest = compute_digest(front_matter, body)

            # Create record
            relpath = str(md_path.relative_to(vault_path))
            rec: FileRec = {
                "id": file_id,
                "relpath": relpath,
                "digest": digest,
                "peers": peers,
                "codebases": codebases,
            }

            index.add_file(rec)

            # If we were looking for a specific file by id
            if limit_file and file_id == limit_file:
                break

        except Exception as e:
            logger.warning(f"Error indexing {md_path}: {e}")
            continue

    return index
