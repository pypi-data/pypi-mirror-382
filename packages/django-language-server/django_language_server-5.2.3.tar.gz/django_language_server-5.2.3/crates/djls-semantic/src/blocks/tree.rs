use djls_source::Span;
use serde::Serialize;

#[salsa::tracked]
pub struct BlockTree<'db> {
    #[returns(ref)]
    pub roots: Vec<BlockId>,
    #[returns(ref)]
    pub blocks: Blocks,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize)]
pub struct BlockId(u32);

impl BlockId {
    pub fn new(id: u32) -> Self {
        Self(id)
    }

    pub fn id(self) -> u32 {
        self.0
    }

    pub fn index(self) -> usize {
        self.0 as usize
    }
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Hash, Serialize)]
pub struct Blocks(Vec<Region>);

impl Blocks {
    pub fn get(&self, id: usize) -> &Region {
        &self.0[id]
    }
}

impl IntoIterator for Blocks {
    type Item = Region;
    type IntoIter = std::vec::IntoIter<Region>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}

impl<'a> IntoIterator for &'a Blocks {
    type Item = &'a Region;
    type IntoIter = std::slice::Iter<'a, Region>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.iter()
    }
}

impl<'a> IntoIterator for &'a mut Blocks {
    type Item = &'a mut Region;
    type IntoIter = std::slice::IterMut<'a, Region>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.iter_mut()
    }
}

impl Blocks {
    pub fn alloc(&mut self, span: Span, parent: Option<BlockId>) -> BlockId {
        let id = BlockId(u32::try_from(self.0.len()).unwrap_or_default());
        self.0.push(Region::new(span, parent));
        id
    }

    pub fn extend_block(&mut self, id: BlockId, span: Span) {
        self.block_mut(id).extend_span(span);
    }

    pub fn set_block_span(&mut self, id: BlockId, span: Span) {
        self.block_mut(id).set_span(span);
    }

    pub fn finalize_block_span(&mut self, id: BlockId, end: u32) {
        let block = self.block_mut(id);
        let start = block.span().start();
        block.set_span(Span::saturating_from_bounds_usize(
            start as usize,
            end as usize,
        ));
    }

    pub fn push_node(&mut self, target: BlockId, node: BlockNode) {
        let span = node.span();
        self.extend_block(target, span);
        self.block_mut(target).nodes.push(node);
    }

    fn block_mut(&mut self, id: BlockId) -> &mut Region {
        let idx = id.index();
        &mut self.0[idx]
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize)]
pub struct Region {
    span: Span,
    nodes: Vec<BlockNode>,
    parent: Option<BlockId>,
}

impl Region {
    fn new(span: Span, parent: Option<BlockId>) -> Self {
        Self {
            span,
            nodes: Vec::new(),
            parent,
        }
    }

    pub fn span(&self) -> &Span {
        &self.span
    }

    pub fn set_span(&mut self, span: Span) {
        self.span = span;
    }

    pub fn nodes(&self) -> &Vec<BlockNode> {
        &self.nodes
    }

    fn extend_span(&mut self, span: Span) {
        let opening = self.span.start().saturating_sub(span.start());
        let closing = span.end().saturating_sub(self.span.end());
        self.span = self.span.expand(opening, closing);
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize)]
pub enum BranchKind {
    Opener,
    Segment,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize)]
pub enum BlockNode {
    Leaf {
        label: String,
        span: Span,
    },
    Branch {
        tag: String,
        marker_span: Span,
        body: BlockId,
        kind: BranchKind,
    },
}

impl BlockNode {
    fn span(&self) -> Span {
        match self {
            BlockNode::Leaf { span, .. } => *span,
            BlockNode::Branch { marker_span, .. } => *marker_span,
        }
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;
    use std::sync::Mutex;

    use camino::Utf8Path;
    use camino::Utf8PathBuf;
    use djls_source::File;
    use djls_source::Span;
    use djls_templates::parse_template;
    use djls_templates::Node;
    use djls_workspace::FileSystem;
    use djls_workspace::InMemoryFileSystem;

    use crate::blocks::grammar::TagIndex;
    use crate::blocks::snapshot::BlockTreeSnapshot;
    use crate::build_block_tree;
    use crate::templatetags::django_builtin_specs;
    use crate::TagSpecs;

    #[salsa::db]
    #[derive(Clone)]
    struct TestDatabase {
        storage: salsa::Storage<Self>,
        fs: Arc<Mutex<InMemoryFileSystem>>,
    }

    impl TestDatabase {
        fn new() -> Self {
            Self {
                storage: salsa::Storage::default(),
                fs: Arc::new(Mutex::new(InMemoryFileSystem::new())),
            }
        }

        fn add_file(&self, path: &str, content: &str) {
            self.fs
                .lock()
                .unwrap()
                .add_file(path.into(), content.to_string());
        }
    }

    #[salsa::db]
    impl salsa::Database for TestDatabase {}

    #[salsa::db]
    impl djls_source::Db for TestDatabase {
        fn create_file(&self, path: &Utf8Path) -> File {
            File::new(self, path.to_owned(), 0)
        }

        fn get_file(&self, _path: &Utf8Path) -> Option<File> {
            None
        }

        fn read_file(&self, path: &Utf8Path) -> std::io::Result<String> {
            self.fs.lock().unwrap().read_to_string(path)
        }
    }

    #[salsa::db]
    impl djls_templates::Db for TestDatabase {}

    #[salsa::db]
    impl crate::Db for TestDatabase {
        fn tag_specs(&self) -> TagSpecs {
            django_builtin_specs()
        }

        fn tag_index(&self) -> TagIndex<'_> {
            TagIndex::from_specs(self)
        }

        fn template_dirs(&self) -> Option<Vec<Utf8PathBuf>> {
            None
        }
    }

    #[test]
    fn test_block_tree_building() {
        let db = TestDatabase::new();

        let source = r"
{% block header %}
    <h1>Title</h1>
{% endblock header %}

{% if user.is_authenticated %}
    <p>Welcome {{ user.name }}</p>
    {% if user.is_superuser %}
        <span>Admin</span>
    {% elif user.is_staff %}
        <span>Manager</span>
    {% else %}
        <span>Regular user</span>
    {% endif %}
{% else %}
    <p>Please log in</p>
{% endif %}

{% for item in items %}
    <li>{{ item }}</li>
{% endfor %}
";

        db.add_file("test.html", source);
        let file = File::new(&db, "test.html".into(), 0);
        let nodelist = parse_template(&db, file).expect("should parse");

        let nodelist_view = {
            #[derive(serde::Serialize)]
            struct NodeListView {
                nodes: Vec<NodeView>,
            }
            #[derive(serde::Serialize)]
            #[serde(tag = "kind")]
            enum NodeView {
                Tag {
                    name: String,
                    bits: Vec<String>,
                    span: Span,
                },
                Variable {
                    var: String,
                    filters: Vec<String>,
                    span: Span,
                },
                Comment {
                    content: String,
                    span: Span,
                },
                Text {
                    span: Span,
                },
                Error {
                    span: Span,
                    full_span: Span,
                    error: String,
                },
            }

            let nodes = nodelist
                .nodelist(&db)
                .iter()
                .map(|n| match n {
                    Node::Tag { name, bits, span } => NodeView::Tag {
                        name: name.clone(),
                        bits: bits.clone(),
                        span: *span,
                    },
                    Node::Variable { var, filters, span } => NodeView::Variable {
                        var: var.clone(),
                        filters: filters.clone(),
                        span: *span,
                    },
                    Node::Comment { content, span } => NodeView::Comment {
                        content: content.clone(),
                        span: *span,
                    },
                    Node::Text { span } => NodeView::Text { span: *span },
                    Node::Error {
                        span,
                        full_span,
                        error,
                    } => NodeView::Error {
                        span: *span,
                        full_span: *full_span,
                        error: error.to_string(),
                    },
                })
                .collect();

            NodeListView { nodes }
        };
        insta::assert_yaml_snapshot!("nodelist", nodelist_view);
        let block_tree = build_block_tree(&db, nodelist);
        insta::assert_yaml_snapshot!("blocktree", BlockTreeSnapshot::from_tree(block_tree, &db));
    }
}
