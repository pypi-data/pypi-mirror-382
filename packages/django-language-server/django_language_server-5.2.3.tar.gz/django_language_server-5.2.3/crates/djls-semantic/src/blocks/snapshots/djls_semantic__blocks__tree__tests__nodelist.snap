---
source: crates/djls-semantic/src/blocks/tree.rs
expression: nodelist_view
---
nodes:
  - kind: Text
    span:
      start: 0
      length: 1
  - kind: Tag
    name: block
    bits:
      - header
    span:
      start: 3
      length: 14
  - kind: Text
    span:
      start: 19
      length: 20
  - kind: Tag
    name: endblock
    bits:
      - header
    span:
      start: 41
      length: 17
  - kind: Text
    span:
      start: 60
      length: 2
  - kind: Tag
    name: if
    bits:
      - user.is_authenticated
    span:
      start: 64
      length: 26
  - kind: Text
    span:
      start: 92
      length: 16
  - kind: Variable
    var: user.name
    filters: []
    span:
      start: 110
      length: 11
  - kind: Text
    span:
      start: 123
      length: 9
  - kind: Tag
    name: if
    bits:
      - user.is_superuser
    span:
      start: 134
      length: 22
  - kind: Text
    span:
      start: 158
      length: 32
  - kind: Tag
    name: elif
    bits:
      - user.is_staff
    span:
      start: 192
      length: 20
  - kind: Text
    span:
      start: 214
      length: 34
  - kind: Tag
    name: else
    bits: []
    span:
      start: 250
      length: 6
  - kind: Text
    span:
      start: 258
      length: 39
  - kind: Tag
    name: endif
    bits: []
    span:
      start: 299
      length: 7
  - kind: Text
    span:
      start: 308
      length: 1
  - kind: Tag
    name: else
    bits: []
    span:
      start: 311
      length: 6
  - kind: Text
    span:
      start: 319
      length: 26
  - kind: Tag
    name: endif
    bits: []
    span:
      start: 347
      length: 7
  - kind: Text
    span:
      start: 356
      length: 2
  - kind: Tag
    name: for
    bits:
      - item
      - in
      - items
    span:
      start: 360
      length: 19
  - kind: Text
    span:
      start: 381
      length: 9
  - kind: Variable
    var: item
    filters: []
    span:
      start: 392
      length: 6
  - kind: Text
    span:
      start: 400
      length: 6
  - kind: Tag
    name: endfor
    bits: []
    span:
      start: 408
      length: 8
  - kind: Text
    span:
      start: 418
      length: 1
